great_expectations/experimental/datasources/dynamic_pandas.py

```diff
@@ -0,0 +1,299 @@
+from __future__ import annotations
+
+import functools
+import inspect
+import logging
+import pathlib
+import re
+from collections import defaultdict
+from pprint import pformat as pf
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Hashable,
+    Iterable,
+    List,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+import pandas as pd
+import pydantic
+
+# https://github.com/pandas-dev/pandas/blob/main/pandas/_typing.py
+from pandas._typing import CompressionOptions, IndexLabel, StorageOptions
+from pydantic import FilePath
+
+# from pydantic.typing import resolve_annotations
+from typing_extensions import Final, Literal, TypeAlias
+
+from great_expectations.experimental.datasources.interfaces import DataAsset
+
+try:
+    from pandas._typing import CSVEngine
+except ImportError:
+    CSVEngine = Literal["c", "python", "pyarrow", "python-fwf"]
+
+logger = logging.getLogger(__file__)
+
+DataFrameFactoryFn: TypeAlias = Callable[..., pd.DataFrame]
+
+# sentinel values
+UNSUPPORTED_TYPE: Final = object()
+
+CAN_HANDLE: Final[Set[str]] = {
+    # builtins
+    "str",
+    "int",
+    "list",
+    "list[str]",
+    "set",
+    "tuple",
+    "dict",
+    "dict[str, str]",
+    "dict[str, list[str]]",
+    "dict[str, Any]",
+    "bool",
+    "None",
+    # typing
+    "Hashable",
+    "Sequence[Hashable]",
+    "Sequence[tuple[int, int]]",
+    "Sequence[Hashable]",
+    "Sequence[str]",
+    "Sequence[int]",
+    "Sequence[tuple[int, int]]",
+    "Literal['infer']",
+    "Literal[False]",
+    "Literal[True]",
+    "Literal['high', 'legacy']",
+    "Literal['frame', 'series']",
+    "Literal['xlrd', 'openpyxl', 'odf', 'pyxlsb']",
+    "Literal[None, 'header', 'footer', 'body', 'all']",
+    "Iterable[object]",
+    "Iterable[Hashable]",
+    # other
+    "Pattern",  # re
+    "Path",  # pathlib
+    "FilePath",  # pydantic
+    # pandas
+    "DtypeArg",
+    "CSVEngine",
+    "IndexLabel",
+    "CompressionOptions",
+    "StorageOptions",
+}
+
+NEED_SPECIAL_HANDLING: Dict[str, Set[str]] = defaultdict(set)
+FIELD_SKIPPED_UNSUPPORTED_TYPE: Set[str] = set()
+FIELD_SKIPPED_NO_ANNOTATION: Set[str] = set()
+
+
+class _SignatureTuple(NamedTuple):
+    name: str
+    signature: inspect.Signature
+
+
+class _FieldSpec(NamedTuple):
+    type: Type
+    default_value: object  # ... for required value
+
+
+@functools.lru_cache(maxsize=64)
+def _replace_builtins(input_: str | type) -> str | type:
+    if not isinstance(input_, str):
+        return input_
+    return input_.replace("list", "List").replace("dict", "Dict")
```

Comment:
This would be faster as complied regex replace.
Was just trying to prove that this was the issue.